File: Binary_Search/LC_33_Search_in_Rotated_Sorted_Array.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.20


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums)

        while l < r:
            mid = l + (r-l)//2
            if nums[mid] == target:
                return mid
            # left part is sorted in ascending order
            if nums[l] < nums[mid]:
                if nums[l] <= target <= nums[mid]:
                    r = mid
                else:
                    l = mid + 1
            # right part is sorted in ascending order
            else:
                if nums[mid] <= target <= nums[r-1]:
                    l = mid + 1
                else:
                    r = mid
        return -1

File: Binary_Search/LC_33_Search_in_Rotated_Sorted_Array.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.20

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size();
        while (l < r){
            int mid = l + (r-l)/2;
            if (nums[mid] == target){
                return mid;
            }
            // left part is sorted in ascending order, so we can easily check if target is in this range
            if (nums[l] < nums[mid]){
                if (nums[l] <= target && target < nums[mid]){
                    r = mid;
                }
                else{
                    l = mid + 1;
                }
            }
            // pivot point is in the left part, so the right part is sorted in ascending order, so we can easily check if target is in this range
            else{
                if (nums[mid] < target && target <= nums[r-1]){
                    l = mid + 1;
                }
                else{
                    r = mid;
                }
            }
        }
        return -1;   
    }
};